### PR TITLE
Standardize performance model outputs

### PR DESCRIPTION
**Pull request by elenya-grant**
# Standardize performance model outputs

Introducing a baseclass for performance models to standardize outputs. The standardized outputs are:
- `commodity_out` in units of `commodity_rate_units`: commodity output profile of length `n_timesteps`
- `total_commodity_produced` in units of `commodity_amount_units`: sum of commodity produced over simulation (adjusted for timestep if necessary)
- `annual_commodity_produced` in units of `commodity_amount_units/yr`: annual commodity production adjusted for a year-long simulation, shape is equal to `plant_life`.
- `replacement_schedule` in units of `unitless`: the percent of the capacity that is replaced per year. Defaults to an array of zeros of length `plant_life`.
- `capacity_factor` in units of `unitless`: the capacity factor of a system as a fraction. Has a shape equal to `plant_life`.
- `rated_commodity_production` in units of `commodity_rate_units`: rated production capacity of the converter. Used to calculate `capacity_factor`
- `operational_life` in units of `yr`: operational life of the technology, defaults to plant life. Tentatively planning on adding it in.

The attributes that each performance model needs to define prior to calling the setup() method in the `PerformanceModelBaseClass` are: `commodity`, `commodity_rate_units` and `commodity_amount_units`. For example, the wind performance models in the setup method would do:
```python
class WindPerformanceBaseClass(PerformanceModelBaseClass):
    def setup(self):
        self.commodity = "electricity"
        self.commodity_rate_units = "kW"
        self.commodity_amount_units = "kW*h"
        super().setup()
```

An electrolyzer performance model would do:
```python
class ElectrolyzerPerformanceBaseClass(PerformanceModelBaseClass):
    def setup(self):
        self.commodity = "hydrogen"
        self.commodity_rate_units = "kg/h"
        self.commodity_amount_units = "kg"
        super().setup()
```

This would resolve some issues around hard-coded tech-specific logic in H2IntegrateModel and the profast finance models, and issues relating to unit convention for varying simulation lengths or timesteps. **To keep this PR small**, this PR will focus on adding these standardized outputs to all the converter models, with the goal of preventing changes in test values. This PR also introduces a lot of TODO notes, which will be addressed in _follow-on PRs_ because most of those changes would likely result in test value changes. Aka - some TODOs will not be removed in this PR. 

This PR is primarily focuses on adding outputs to performance models and adding basic tests to ensure that all the outputs are set (not testing the value that they're set to, just that they're not zero). To ensure that the planned changes that may result in different test values or larger framework changes to core models (which this PR lays the foundation for) are highlighted, these are planned to be done in future PRs so that all the additions in this PR do not distract from those changes.

Out of scope for this PR is:
- standardizing performance model inputs (not planned as part of this larger effort)
- adding in technology-specific unit tests to test specific output values (captured in Issue #481)
- making technology models flexible to varying timesteps or simulation lengths
- removing existing outputs of performance models that are equivalent to the new outputs introduced with the `PerformanceModelBaseClass`

**The follow-on PRs will be**:
- Update electrolyzer outputs to be per-year (this will change a lot of test values that are testing hydrogen production, LCOH, or any LCO values of components downstream of the electrolyzer). This could be part of the PR for Issue #485
- Update storage/control models to use the naming "commodity_rate_units" and "commodity_amount_units" instead of "commodity_units" or "commodity_storage_units". This is noted in Issue #486
- Update finance models (this will also change test result values). This is noted in Issue #485. The updates are:
   - use capacity factor as the long term utilization
   - use rated commodity production as the capacity
 - Remove tech-specific logic in H2IntegrateModel, remove ElectricitySumComp, remove logic in "generic_summer.py". This would be part of the PR to resolve issue #486
 - Update feedstock component to use commodity and commodity rate units (This would be part of the PR to resolve issue #486 
 - Remove tech-specific or commodity-specific hard-coded logic in core models, this is noted in Issue #488
 
These PRs can get started once this PR is merged in. 

Benefits of this PR are:
- standardized output naming can reduce tech-specific hard-coded logic in H2IntegrateModel, finance models, etc
- standardized output naming may also reduce the number of places in the code that a new commodity type has to be added when a tech with a new commodity type is added
- standardized usage of rate units vs amount units

## Section 1: Type of Contribution

- [ ] Feature Enhancement
  - [x] Framework
  - [ ] New Model
  - [ ] Updated Model
  - [ ] Tools/Utilities
  - [ ] Other (please describe):
- [ ] Bug Fix
- [ ] Documentation Update
- [ ] CI Changes
- [ ] Other (please describe):

## Section 2: Draft PR Checklist

- [x] Open draft PR
- [x] Describe the feature that will be added
- [x] Fill out TODO list steps
- [x] Describe requested feedback from reviewers on draft PR
- [ ] Complete Section 7: New Model Checklist (if applicable)

### TODO:
- [ ] Update/add outputs to converter models and add basic unit-tests that outputs are set

   - [x] wind 
   - [x] solar 
   - [x] water_power 
   - [x] ammonia
      - [x] simple ammonia 
      - [x] ammonia synloop
   - [x] co2 (in progress/partially done, has remaining todos for removing duplicate outputs)
      - [x] doc
      - [x] oae 
   - [x] grid
   - [x] hopp
   - [x] hydrogen (electrolyzer) (in progress/partially done, has remaining todos)
   - [x] hydrogen (geologic) 
   - [x] iron
      - [ ] iron mine (old - should be deleted)
      - [x] standalone iron mine (Martin mine)
      - [ ] iron wrapper (old - should be deleted)
      - [ ] iron plant (old - should be deleted) 
      - [x] iron dri
   - [x] methanol 
      - [x] smr
      - [x] co2h 
   - [x] natural gas 
   - [x] steel 
      - [x] steel.py (used in Ex 1)
      - [ ] electric arc funance (old - should be deleted)
      - [x] steel eaf plant 
   - [x] water (desal)
   - [ ] paper mill in Example 6 (**future PR**)
   - [x] nitrogen ASU model
   
- [x] Update/add outputs to storage models 

   - [-] hydrogen (N/A)
   - [x] battery 
   - [ ] `simple_generic_storage` (**future PR**, Issue #486)
   - [ ] `simple_storage_autosizing` (**future PR**, Issue #486)
   
- [-] Update/add outputs to feedstock model and fix units (**future PR**)
- [-] Update combiners and splitters (as necessary, **mostly done in future PR**)
- [-] Update ProFAST finance models to use capacity factor as utilization and `rated_commodity_production` as capacity (**make as a separate PR**)
- [-] Update `AdjustedCapexOpexComp` if needed.
- [x] Update/fix tests as needed
- [x] Update post-processing functions as necessary 
- [-] Update documentation on adding a new technology (**future PR**)
- [x] Add unit-tests to check that all outputs set via `PerformanceModelBaseClass` are given values in  the `compute()` method of parent classes.

### Type of Reviewer Feedback Requested (on Draft PR)

**Structural feedback**:
- thoughts on standardized output naming convention?

**Implementation feedback**:
- Changes to the wind, solar, water power, and electrolyzer converters have already been done and can be reviewed now for reviewers to provide high-level feedback. 

**Other feedback**:
- do we want to assume a year is 8760 hours or use what OpenMDAO assumes is a year (8765.812776 hours/year)?

## Section 3: General PR Checklist

- [x] PR description thoroughly describes the new feature, bug fix, etc.
- [x] Added tests for new functionality or bug fixes
- [x] Tests pass (If not, and this is expected, please elaborate in the Section 6: Test Results)
- [x] Documentation
  - [x] Docstrings are up-to-date
  - [x] Related `docs/` files are up-to-date, or added when necessary
  - [x] Documentation has been rebuilt successfully
  - [-] Examples have been updated (if applicable)
- [ ] `CHANGELOG.md` has been updated to describe the changes made in this PR

### Existing Related Issues
Units for varying timesteps and simulation lengths: Issue #244, #204, and #387 (may be partially resolved with this PR)
Standardized naming conventions: Issue #223 (this would be partially resolved with this PR)
Remove dependence on name of the technologies in H2I: Issue #374 (would be partially/fully resolved in this PR)
Issue about converter baseclass (somewhat related): Issue #231

### New issues made
- Issue #481 (tests that are missing or need to be reformatted)
- Issue #485 (for follow-on PR)
- Issue #486 (for follow-on PR)
- Issue #487 (update docs on how to add a new tech)
- Issue #488 (for follow-on PR)

### Section 4.1: New Files
- `h2integrate/converters/hydrogen/test/test_pem_electrolyzer_performance.py`
- `h2integrate/converters/steel/test/test_simple_steel.py`
- `h2integrate/converters/methanol/test/test_methanol.py`
- `h2integrate/converters/water_power/test/test_hydro_power.py`

### Section 4.2: Modified Files
Most converter performance models or performance baseclasses and some variable names in subtests

## Section 5: Additional Supporting Information

Future development (in other PRs than the ones noted above) that could build on this framework are:
- standardize outputs for techs that produce multiple commodities
- standardize inputs and outputs for techs that require/use feedstocks
- updates to specific technology models to account for varying timestep (`dt`) or simulation length (`n_timesteps`)

## Section 6: Test Results, if applicable

No test values changed!

### CODE REVIEW

File: h2integrate/converters/ammonia/test/test_ammonia_synloop_model.py [hunk 1]
```diff
@@ -40,13 +42,122 @@ def make_synloop_config():
...
+
+    prob.run_model()
+
+    commodity = "ammonia"
+    commodity_amount_units = "kg"
+    commodity_rate_units = "kg/h"
+    plant_life = int(plant_config["plant"]["plant_life"])
+    n_timesteps = int(plant_config["plant"]["simulation"]["n_timesteps"])
+
+    # Check that replacement schedule is between 0 and 1
```
Comment (bayc) [inline h2integrate/converters/ammonia/test/test_ammonia_synloop_model.py]: These type of tests are useful! Wondering if we also have regression tests for this tech? I see the test below is checking some values, but wondering about checking the values of these standardized outputs. Is that for a future PR? Definitely have a lot of thoughts on testing.

File: h2integrate/core/model_baseclasses.py [hunk 2]
```diff
@@ -382,3 +382,83 @@ def compute(self, inputs, outputs, discrete_inputs, discrete_outputs):
...
+            shape=self.plant_life,
+            units=f"({self.commodity_amount_units})/year",
+        )
+        # lifetime estimate of item replacements, represented as a fraction of the capacity.
+        self.add_output("replacement_schedule", val=0.0, shape=self.plant_life, units="unitless")
+        # capacity factor is the ratio of actual production / maximum production possible
+        self.add_output(
+            "capacity_factor",
+            val=0.0,
+            shape=self.plant_life,
```
[the remark this answers is not in the document]
Comment (elenya-grant) [reply h2integrate/core/model_baseclasses.py]: I think that's a valid name choice, I'd want to get the rest of the teams opinion on it before changing it. Could also be done in a separate PR

File: h2integrate/converters/ammonia/ammonia_synloop.py [hunk 3]
```diff
@@ -539,7 +548,7 @@ def compute(self, inputs, outputs, discrete_inputs, discrete_outputs):
         o2_price_base = self.config.oxygen_price_base  # USD / kg O2
 
         # Get total production/consumption
-        nh3_prod = inputs["total_ammonia_produced"]  # kg NH3 /year
+        nh3_prod = inputs["annual_ammonia_produced"].mean()  # kg NH3 /year
```
[the remark this answers is not in the document]
Comment (elenya-grant) [reply h2integrate/converters/ammonia/ammonia_synloop.py]: Only the electrolyzer model and the geoh2 models account for lifetime performance, and only some models (I'm not sure which ones) are/may be flexible to different simulation period or timesteps. For most models, the annual production is the sum of the production over the simulation (because most models assume a year-long simulation). This is so that models that DO account for lifetime performance can have that lifetime performance reflected in the finances. A future PR will include that update

File: h2integrate/converters/ammonia/ammonia_synloop.py [hunk 4]
```diff
@@ -140,6 +140,9 @@ class AmmoniaSynLoopPerformanceModel(ResizeablePerformanceModelBaseClass):
     """
 
     def setup(self):
+        self.commodity = "ammonia"
```
[the remark this answers is not in the document]
Comment (elenya-grant) [reply h2integrate/converters/ammonia/ammonia_synloop.py]: Since this was mentioned by multiple folks, I will say that I think this should be discussed with the larger team at some point but could be done after the planned development related to this PR is complete.

File: h2integrate/converters/wind/wind_plant_baseclass.py [hunk 5]
```diff
@@ -1,21 +1,13 @@
...
-        n_timesteps = self.options["plant_config"]["plant"]["simulation"]["n_timesteps"]
-        self.add_output(
-            "electricity_out",
-            val=0.0,
-            shape=n_timesteps,
-            units="kW",
-            desc="Power output from WindPlant",
-        )
+        self.commodity = "electricity"
+        self.commodity_rate_units = "kW"
```
[the remark this answers is not in the document]
Comment (jaredthomas68) [reply h2integrate/converters/wind/wind_plant_baseclass.py]: I see your point. I think this is related to #365, so tagging here so the comments are not lost.

**Review by jaredthomas68 — CHANGES_REQUESTED** [review]
Looking good. I still have a few outstanding comments I would like to get resolved prior to merging this in.

File: h2integrate/converters/hydrogen/electrolyzer_baseclass.py [hunk 6]
```diff
@@ -33,7 +53,9 @@ class ElectrolyzerCostBaseClass(CostModelBaseClass):
     def setup(self):
         n_timesteps = self.options["plant_config"]["plant"]["simulation"]["n_timesteps"]
         super().setup()
-        self.add_input("total_hydrogen_produced", val=0.0, units="kg/year")
+        self.add_input(
+            "total_hydrogen_produced", val=0.0, units="kg"
+        )  # NOTE: unsure if this is used
```
[the remark this answers is not in the document]
Comment (elenya-grant) [reply h2integrate/converters/hydrogen/electrolyzer_baseclass.py]: That's only an input to the electrolyzer cost model base class (not performance model), but it doesnt appear to be used in any of the cost models. I'll remove it.

File: h2integrate/converters/ammonia/ammonia_synloop.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (elenya-grant) [reply h2integrate/converters/ammonia/ammonia_synloop.py]: that's going to be a separate PR because it'll change a lot of the test results. In the separate PR - I was only going to change the logic for the electrolyzer so that change is isolated.

**Comment by johnjasa** [comment]
> @johnjasa
> 
> I realized that to remove some tech-specific logic in `H2IntegrateModel`, that it would require setting the attributes `commodity`, `commodity_rate_units`, and `commodity_amount_units` in the `initialize()` method instead of the `setup()` method. What are you thoughts on this:
> 
> ```python
> class WindPerformanceBaseClass(PerformanceModelBaseClass):
>     def initialize(self):
>         super().initialize()
>         self.commodity = "electricity"
>         self.commodity_rate_units = "kW"
>         self.commodity_amount_units = "kW*h"
> ```
> 
> instead of this:
> 
> ```python
>     def setup(self):
>         self.commodity = "electricity"
>         self.commodity_rate_units = "kW"
>         self.commodity_amount_units = "kW*h"       
>         super().setup()
> ```
> 
> This would also be okay in tech-neutral technologies because most of those would have to define the commodity and units in the `tech_config` which we could also access within `H2IntegrateModel`

Thanks Elenya for this idea! We just chatted and it seems worthwhile to implement and would lead to smoother code modification down the line.

I'll sit tight as you update things then will re-review.
